adr/ADR-44.md

```diff
@@ -0,0 +1,82 @@
+# Versioning for JetStream Assets
+
+| Metadata | Value             |
+|----------|-------------------|
+| Date     | 2024-07-22        |
+| Author   | @ripienaar        |
+| Status   | Approved          |
+| Tags     | jetstream, server |
+
+# Context and Problem Statement
+
+As development of the JetStream feature progress there is a complex relationship between connected-server, JetStream 
+meta-leader and JetStream asset host versions that requires careful consideration to maintain compatibility.
+
+ * The server a client connects to might be a Leafnode on a significantly older release and might not even have 
+   JetStream enabled
+ * The meta-leader could be on a version that does not support a feature and so would not forward the API fields it 
+   is unaware of to the assigned servers
+ * The server hosting an asset might not match the meta-leader and so can't honor the request for a certain 
+   configuration and would silently drop fields
+
+In general our stance is to insist on homogenous cluster versions in the general case but it's not reasonable to expect this
+for leafnode servers nor is it possible during upgrading and other maintenance windows.
+
+Our current approach is to check the connected-server version and project that it's representitive of the cluster as 
+a whole but this is a known incorrect approach especially for Leafnodes as mentioned above.
+
+This ADR specifies a way for the servers to expose some versioning information to help clients improve the 
+compatability story.
+
+# Server-set metadata
+
+In lieu of enabling fine grained API versioning we want to start thinking about asset versioning instead, the server
+should report the version that created an asset and it should report the version that is hosting the asset and 
+potentially expose a asset version.
+
+We'll store this in the existing `metadata` field allowing us to expand this in time, today we propose the following:
+
+| Name                           | Description                                      |
+|--------------------------------|--------------------------------------------------|
+| `_nats.server.version`         | The current server version hosting an asset      |
+| `_nats.server.created.version` | The version server that first created this asset |
+| `_nats.server.created.time`    | The time the asset was first created             |
```

Comment:
future here:

```
_nats.client.created.version
_nats.client.created.lang
_nats.client.created.tool (optional for cli, terraform, nack etc)
```

And a similar set for updated maybe, I wouldn't want to keep a full update history here somehow just most recent updated lang/tool/version

Reply:
We should make a note that we reserve `_nats` or `nats` in headers and metadata fields. If set by an API call the server should possibly error or reject the update.

Reply:
Already in ADR-33 but we need to communicate while wider